resdk/data_upload/reads.py

```diff
@@ -0,0 +1,108 @@
+        resdk_logger.start_logging()
+
+    # Read and validate the annotation template
+    basedir = os.path.dirname(os.path.abspath(samplesheet_path))
+    if verbose:
+        print(
+            "\nChecking the sample annotation spreadsheet at:"
+            "\n{}".format(samplesheet_path)
+        )
+    samplesheet = FileImporter(samplesheet_path)
+    err_samples = samplesheet.invalid_samples
+
+    for sample in samplesheet.sample_list:
+        reads = None
+        try:
+            _validate_upload(collection, sample, err_samples)
+
+            # Upload the reads
+            path = _parse_paths(basedir, sample.path)
+            path2 = _parse_paths(basedir, sample.path)
+            if path and path2:
+                print('Uploading data for the sample: {}'.format(sample.name))
+                reads = resolwe.run('upload-fastq-paired',
+                                    input={'src1': path, 'src2': path2},
+                                    collections=[collection])
+            elif path and not path2:
+                print('Uploading data for the sample: {}'.format(sample.name))
+                reads = resolwe.run('upload-fastq-single',
+                                    input={'src': path},
+                                    collections=[collection])
+
+        except (FileNotFoundError, ValueError) as ex:
+            if verbose:
+                print(ex)
+            err_samples.add(sample.name)
+
+        if reads:
+            # Create sample object with appropriate name
+            reads.sample.delete(force=True)
+            main_sample = resolwe.sample.create(name=sample.name)
+            main_sample.add_data(reads)
+            main_sample.save()
+            collection.add_samples(main_sample)
+
+    # Report failures
+    print(
+        "The following samples could not be uploaded: "
+        "{}.".format(', '.join(err_samples))
+    )
+
+    # Annotate the samples
+    collection.annotate(samplesheet, source_type='upload', verbose=False)
+
+
+def _parse_paths(base, filepath):
+    """Convert filepaths into list."""
+    if filepath:
+        return [os.path.join(base, s) for s in filepath.split(',')]
+    else:
+        return []
+
+
+def _validate_upload(collection, sample, pre_invalid):
+    """Check if the sample is valid to upload to this collection."""
+    # Check if the data is already in the collection
+    skip_msg = "Skipping upload of '{}': ".format(sample.name)
+    try:
+        collection.data.get(name=sample.name)
+        raise ValueError(skip_msg + "File already uploaded.")
+    except LookupError:
+        pass
+
+    # Don't upload if the annotation info is invalid
+    if sample.name in pre_invalid:
+        raise ValueError(skip_msg + "Invalid annotation.")
+
+    # Check if filepaths were given
+    if not sample.path:
+        raise ValueError(skip_msg + "No forward reads given.")
+
+    # Check if at least one file has the proper filetype
+    validtypes = ['.fq', '.fastq']
```

Comment:
This is duplicated form resolwe-bio. Fetch it from the api instead.

Reply:
Do we have a built-in function to do this? I can see how this information is stored using the `validate_regex` key on the various elements of `Process.input_schema`, but I'm not sure of the best way to parse this.